server/controller/src/main/java/ai/starwhale/mlops/domain/model/ModelService.java

```diff
@@ -371,56 +378,131 @@ public void upload(MultipartFile dsFile,
+            RevertManager.create(bundleManager, modelDao)
+                    .revertVersionTo(modelVersionEntity.getModelId(), modelVersionEntity.getId());
+        }
+        return ModelUploadResult.builder()
+                .uploadId(modelVersionEntity.getId().toString())
+                .existed(existed)
+                .build();
+    }
+
+    @Transactional
+    public void uploadModel(Long modelVersionId, String signature,
+                            MultipartFile modelFile, ModelUploadRequest uploadRequest) {
+        ModelVersionEntity modelVersionEntity = modelVersionMapper.find(modelVersionId);
+        if (modelVersionEntity == null
+                || Objects.equals(modelVersionEntity.getStatus(), ModelVersionEntity.STATUS_AVAILABLE)) {
+            throw new StarwhaleApiException(
+                    new SwValidationException(ValidSubject.MODEL),
+                    HttpStatus.BAD_REQUEST
+            );
+        }
+        String modelPath = storagePathCoordinator.allocateCommonModelPoolPath(
+                uploadRequest.getProject(), signature);
+
+        try {
+            storageAccessService.put(modelPath, modelFile.getInputStream());
+        } catch (IOException e) {
+            log.error("upload model failed {}", uploadRequest.getSwmp(), e);
+            throw new StarwhaleApiException(
+                    new SwProcessException(ErrorType.STORAGE),
+                    HttpStatus.INTERNAL_SERVER_ERROR
+            );
+        }
+    }
+
+    @Transactional
+    public void uploadSrc(Long modelVersionId, MultipartFile multipartFile, ModelUploadRequest uploadRequest) {
+        ModelVersionEntity modelVersionEntity = modelVersionMapper.find(modelVersionId);
+        if (modelVersionEntity == null
+                || Objects.equals(modelVersionEntity.getStatus(), ModelVersionEntity.STATUS_AVAILABLE)) {
+            throw new StarwhaleApiException(
+                    new SwValidationException(ValidSubject.MODEL),
+                    HttpStatus.BAD_REQUEST
+            );
+        }
+        //upload to storage
+        final String storagePath = modelVersionEntity.getStoragePath();
         String jobContent = "";
-        try (final InputStream inputStream = dsFile.getInputStream()) {
+        try (final InputStream inputStream = multipartFile.getInputStream()) {
             // only extract the eval job file content
             // TODO: replace with oss path content
             // but update only for job
             jobContent = new String(
                     Objects.requireNonNull(
-                            TarFileUtil.getContentFromTarFile(dsFile.getInputStream(), "src", "eval_jobs.yaml")));
+                            TarFileUtil.getContentFromTarFile(
+                                multipartFile.getInputStream(), "src", "eval_jobs.yaml")
+                    )
+            );
             TarFileUtil.extract(inputStream, (name, size, in) ->
```

Comment:
is the src still a tar file?

Reply:
yes, there has three types: _manifest.yaml, src.tar(code and config yamls), models, so compress the src files as one tar when upload can reduce upload times

Reply:
So what is the initial requirement for splitting model from tar to dir? For versioning , disk saving for server or something else?

Reply:
‘support duplicate validation for model file to avoid multi upload the same model’